### PR TITLE
Cron scheduler

### PR DESCRIPTION
**Pull request by ikurian-coatue**
Cron based scheduler for airflow (better_airflow)

- added a cron based scheduler
- updated execution date to be wall time vs interval time
- added a heartbeat table to handle recovery if/when scheduler is down

Tests
- [x] Schedules jobs based on cron interval
- [x] updates heartbeat per dag to track last time scheduler was run
- [x] verify that only latest schedule change is acted on
- test recovery logic on 
     - [x] relative dags ex: "* * * * *"
     - [x] daily dags
     - [x] weekly dags
     - [x] dags with schedule changes

### CODE REVIEW

File: tests/test_cron_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
+from __future__ import absolute_import
```
Comment (pathikrit) [inline tests/test_cron_scheduler.py]: file an issue to make this part of CI

File: airflow/models/__init__.py [hunk 2]
```diff
@@ -3238,20 +3238,29 @@ def following_schedule(self, dttm):
...
             # we don't want to rely on the transitions created by
             # croniter as they are not always correct
+            self.log.info(f"In following_schedule passed in dttm={dttm}")
             dttm = pendulum.instance(dttm)
-            naive = timezone.make_naive(dttm, self.timezone)
+            self.log.info(f"In following_schedule updated  dttm={dttm}")
+            if not tz:
+                naive = timezone.make_naive(dttm, self.timezone)
+            else:
+                naive = timezone.make_naive(dttm, tz)
```
Comment (pathikrit) [inline airflow/models/__init__.py]: can you just do:
```
                naive = timezone.make_naive(dttm, tz or self.timezone)
```

File: airflow/jobs.py [hunk 3]
```diff
@@ -775,6 +777,96 @@ def update_import_errors(session, dagbag):
...
+        last_heartbeat = SchedulerHeartbeat.get(dag.dag_id)
+        self.log.info(f"Last heartbeat: {last_heartbeat}")
+        window_start = last_heartbeat.last_heartbeat if last_heartbeat else min_window_start
+        # convert to local timezone since this is what we intuitively schedule our jobs with
+        window_start = local_tz.convert(window_start)
+
+        if  window_start < max_window_start:
+            window_start = max_window_start
+        elif window_start > min_window_start:
+            window_start = min_window_start
```
Comment (pathikrit) [inline airflow/jobs.py]: one liner:
```
window_start = min(max_window_start, window_start)
window_start = max(min_window_start, window_start)
```

File: airflow/jobs.py [hunk 4]
```diff
@@ -775,6 +777,96 @@ def update_import_errors(session, dagbag):
...
+        min_interval = 60 # 1 min
+        min_window_start = window_end - timedelta(seconds=min_interval)
+
+        last_heartbeat = SchedulerHeartbeat.get(dag.dag_id)
+        self.log.info(f"Last heartbeat: {last_heartbeat}")
+        window_start = last_heartbeat.last_heartbeat if last_heartbeat else min_window_start
+        # convert to local timezone since this is what we intuitively schedule our jobs with
+        window_start = local_tz.convert(window_start)
+
+        if  window_start < max_window_start:
```
Comment (pathikrit) [inline airflow/jobs.py]: don't we have to update the window end too?

File: airflow/jobs.py [hunk 5]
```diff
@@ -775,6 +777,96 @@ def update_import_errors(session, dagbag):
...
+    def get_recovery_window(self, dag, session):
+        """
+        Determines a window to look at to determine if a job has run. In the event of a recovery from
+        deploy or crash, we have established upper and lower bounds for the window, 30 min and 1 min respectively.
+        :param dag: the dag being acted on
+        :param session: sqlalchemy session
+        :return: a tuple of the start and end window datetimes
+        """
+        window_end = timezone.utcnow()
+        local_tz = pendulum.timezone("America/New_York")
```
Comment (pathikrit) [inline airflow/jobs.py]: make this a constant in Config.ini?

File: airflow/jobs.py [hunk 6]
```diff
@@ -775,6 +777,96 @@ def update_import_errors(session, dagbag):
...
+        # convert to local timezone since this is what we intuitively schedule our jobs with
+        window_end = local_tz.convert(window_end)
+
+        # upper bound for the window
+        max_interval = 30 * 60 # 30 min
```
Comment (pathikrit) [inline airflow/jobs.py]: these should all be constants in ini file IMO

File: tests/test_cron_scheduler.py (hunk 1 above)
Comment (vdat-coatue) [reply tests/test_cron_scheduler.py]: https://github.com/coatue/airflow/issues/124

File: airflow/jobs.py (hunk 4 above)
Comment (vdat-coatue) [reply airflow/jobs.py]: window end is `now()` so we don't need to update it, as we only want the window interval to be between some x minutes in the past to now(). But will add comments to code

**Pull request merged**